pkg/watch/watch.go

```diff
@@ -178,34 +187,43 @@ func (o *WatchClient) WatchAndPush(out io.Writer, parameters WatchParameters) er
 	// mutex as they are shared between goroutines to communicate
 	// sync state/events.
 	var (
-		changeLock   sync.Mutex
-		dirty        bool
-		lastChange   time.Time
-		watchError   error
-		deletedPaths []string
-		changedFiles []string
+		lastChange                     time.Time
+		watchError                     error
+		deletedPaths                   []string
+		changedFiles                   []string
+		hasFirstSuccessfulPushOccurred bool
+		wg                             sync.WaitGroup
+		showWaitingMessage             bool
+		delay                          time.Duration
 	)
 
+	delay = 1 * time.Second
+
 	watcher, err := fsnotify.NewWatcher()
 	if err != nil {
 		return fmt.Errorf("error setting up filesystem watcher: %v", err)
 	}
 	defer watcher.Close()
 	defer close(parameters.ExtChan)
 
+	// adding watch on the root folder and the sub folders recursively
+	// so directory and the path in addRecursiveWatch() are the same
+	err = addRecursiveWatch(watcher, parameters.Path, parameters.FileIgnores)
+	if err != nil {
+		return fmt.Errorf("error watching source path %s: %v", parameters.Path, err)
+	}
+
+	printInfoMessage(out, parameters.Path)
 	// This goroutine listens for either file change events from fsnotify, fs errors, or a terminate signal
 	// The results are stored in the variables defined in the var( ... ) block above
-	go func() {
-		for {
-			select {
-			case extMsg := <-parameters.ExtChan:
-				if extMsg {
-					changeLock.Lock()
-					watchError = ErrUserRequestedWatchExit
-					changeLock.Unlock()
-				}
-			case event := <-watcher.Events:
-				changeLock.Lock()
+	wg.Add(1)
+	for {
+		select {
+		case event := <-watcher.Events:
+			// check if the delay between events is more than a second; if not don't push again
+			if time.Now().After(lastChange.Add(delay)) {
```

Comment:
I don't think it will work this way. Instead, you should create a timer of 'delay' duration that you start when an event is received, and you add a case to catch the end of the timer. If you receive another event before the end of the timer, you reinitialize the timer to delay, and if you receive a timer end, you can proceed

Reply:
PTAL at 86c8e04.